storm-core/src/jvm/org/apache/storm/metric/MetricsConsumerBolt.java

```diff
@@ -47,17 +68,71 @@ public void prepare(Map stormConf, TopologyContext context, OutputCollector coll
         }
         _metricsConsumer.prepare(stormConf, _registrationArgument, context, collector);
         _collector = collector;
+        _taskExecuteThread = new Thread(new MetricsHandlerRunnable());
+        _taskExecuteThread.setDaemon(true);
+        _taskExecuteThread.start();
     }
     
     @Override
     public void execute(Tuple input) {
-        _metricsConsumer.handleDataPoints((IMetricsConsumer.TaskInfo)input.getValue(0), (Collection)input.getValue(1));
+        // remove older tasks if task queue exceeds the max size
+        if (_taskQueue.size() > _maxRetainMetricTuples) {
+            while (_taskQueue.size() - 1 > _maxRetainMetricTuples) {
```

Comment:
I guess maybe it could drop one more item but it's not a big deal anyway. Ideally, taskQueue exceeding the max retain metric tuples should not occur.